File: packages/server/src/routes/files.ts
import { Hono } from "hono";
import type { Database } from "bun:sqlite";
import { boardsDb } from "@glue-paste-dev/core";
import type { BoardId } from "@glue-paste-dev/core";
import { readdirSync, statSync, mkdirSync, rmSync } from "fs";
import { resolve, join, relative, basename } from "path";

const IGNORED = new Set([
  ".git",
  "node_modules",
  ".next",
  "dist",
  "build",
  "__pycache__",
  ".venv",
  ".DS_Store",
  ".glue-paste",
]);

const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB

function sanitizeFilename(name: string): string {
  return basename(name).replace(/[/\\]/g, "_").replace(/\0/g, "_");
}

interface FileEntry {
  name: string;
  type: "file" | "directory";
  path: string;
}

export function fileRoutes(db: Database) {
  const app = new Hono();

  // GET /api/files/board/:boardId?path=<subdir>
  app.get("/board/:boardId", (c) => {
    const boardId = c.req.param("boardId") as BoardId;
    const board = boardsDb.getBoard(db, boardId);
    if (!board) {
      return c.json({ error: "Board not found" }, 404);
    }

    const subPath = c.req.query("path") || "";
    const rootDir = resolve(board.directory);
    const targetDir = resolve(rootDir, subPath);

    // Prevent directory traversal
    if (!targetDir.startsWith(rootDir)) {
      return c.json({ error: "Invalid path" }, 400);
    }

    try {
      const entries = readdirSync(targetDir);
      const result: FileEntry[] = [];

      for (const name of entries) {
        if (IGNORED.has(name)) continue;

        try {
          const fullPath = join(targetDir, name);
          const stat = statSync(fullPath);
          const relPath = relative(rootDir, fullPath);
          result.push({
            name,
            type: stat.isDirectory() ? "directory" : "file",
            path: relPath,
          });
        } catch {
          // Skip entries we can't stat
        }
      }

      // Sort: directories first, then alphabetically
      result.sort((a, b) => {
        if (a.type !== b.type) return a.type === "directory" ? -1 : 1;
        return a.name.localeCompare(b.name);
      });

      return c.json(result);
    } catch {
      return c.json({ error: "Cannot read directory" }, 400);
    }
  });

  // GET /api/files/board/:boardId/tree — recursive file tree for fuzzy search
  app.get("/board/:boardId/tree", (c) => {
    const boardId = c.req.param("boardId") as BoardId;
    const board = boardsDb.getBoard(db, boardId);
    if (!board) {
      return c.json({ error: "Board not found" }, 404);
    }

    const rootDir = resolve(board.directory);
    const maxFiles = 10000;
    const maxDepth = 20;
    const entries: FileEntry[] = [];
    let truncated = false;

    function walk(dir: string, depth: number) {
      if (depth > maxDepth || truncated) return;
      let names: string[];
      try {
        names = readdirSync(dir);
      } catch {
        return;
      }
      for (const name of names) {
        if (IGNORED.has(name)) continue;
        if (entries.length >= maxFiles) {
          truncated = true;
          return;
        }
        try {
          const fullPath = join(dir, name);
          const stat = statSync(fullPath);
          const relPath = relative(rootDir, fullPath);
          const isDir = stat.isDirectory();
          entries.push({ name, type: isDir ? "directory" : "file", path: relPath });
          if (isDir) walk(fullPath, depth + 1);
        } catch {
          // skip unreadable
        }
      }
    }

    walk(rootDir, 0);
    return c.json({ entries, truncated });
  });

  // POST /api/files/board/:boardId/upload/:cardId
  app.post("/board/:boardId/upload/:cardId", async (c) => {
    const boardId = c.req.param("boardId") as BoardId;
    const cardId = c.req.param("cardId");
    const board = boardsDb.getBoard(db, boardId);
    if (!board) {
      return c.json({ error: "Board not found" }, 404);
    }

    const rootDir = resolve(board.directory);
    const attachmentsDir = join(rootDir, ".glue-paste", "attachments", cardId);
    mkdirSync(attachmentsDir, { recursive: true });

    const body = await c.req.parseBody({ all: true });
    const rawFiles = body["files"];
    const files: File[] = Array.isArray(rawFiles)
      ? (rawFiles.filter((f) => f instanceof File) as File[])
      : rawFiles instanceof File
        ? [rawFiles]
        : [];

    if (files.length === 0) {
      return c.json({ error: "No files provided" }, 400);
    }

    const savedPaths: string[] = [];

    for (const file of files) {
      if (file.size > MAX_FILE_SIZE) {
        return c.json({ error: `File "${file.name}" exceeds 10MB limit` }, 400);
      }
      const safeName = sanitizeFilename(file.name) || "upload";
      const destPath = join(attachmentsDir, safeName);
      const buffer = await file.arrayBuffer();
      await Bun.write(destPath, buffer);
      savedPaths.push(relative(rootDir, destPath));
    }

    return c.json(savedPaths);
  });

  // GET /api/files/board/:boardId/attachments/:cardId
  app.get("/board/:boardId/attachments/:cardId", (c) => {
    const boardId = c.req.param("boardId") as BoardId;
    const cardId = c.req.param("cardId");
    const board = boardsDb.getBoard(db, boardId);
    if (!board) {
      return c.json({ error: "Board not found" }, 404);
    }

    const rootDir = resolve(board.directory);
    const attachmentsDir = join(rootDir, ".glue-paste", "attachments", cardId);

    try {
      const names = readdirSync(attachmentsDir);
      const files = names.map((name) => relative(rootDir, join(attachmentsDir, name)));
      return c.json(files);
    } catch {
      return c.json([]);
    }
  });

  // DELETE /api/files/board/:boardId/attachments/:cardId/:filename
  app.delete("/board/:boardId/attachments/:cardId/:filename", (c) => {
    const boardId = c.req.param("boardId") as BoardId;
    const cardId = c.req.param("cardId");
    const filename = c.req.param("filename");
    const board = boardsDb.getBoard(db, boardId);
    if (!board) {
      return c.json({ error: "Board not found" }, 404);
    }

    const safeName = sanitizeFilename(filename);
    if (!safeName) {
      return c.json({ error: "Invalid filename" }, 400);
    }

    const rootDir = resolve(board.directory);
    const filePath = join(rootDir, ".glue-paste", "attachments", cardId, safeName);

    try {
      rmSync(filePath);
    } catch {
      // file may not exist
    }

    return c.json({ ok: true });
  });

  // DELETE /api/files/board/:boardId/attachments/:cardId
  app.delete("/board/:boardId/attachments/:cardId", (c) => {
    const boardId = c.req.param("boardId") as BoardId;
    const cardId = c.req.param("cardId");
    const board = boardsDb.getBoard(db, boardId);
    if (!board) {
      return c.json({ error: "Board not found" }, 404);
    }

    const rootDir = resolve(board.directory);
    const attachmentsDir = join(rootDir, ".glue-paste", "attachments", cardId);

    try {
      rmSync(attachmentsDir, { recursive: true, force: true });
    } catch {
      // ignore if directory doesn't exist
    }

    return c.json({ ok: true });
  });

  return app;
}

File: rust/crates/tauri-app/src/commands/files.rs
/// File commands — browse board directory, manage card attachments.
/// Upload (multipart) stays on HTTP; everything else is wired here.
use tauri::State;
use glue_paste_dev_core::db::boards;
use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};
use crate::state::AppState;

fn map_err(e: impl std::fmt::Display) -> String { e.to_string() }

const IGNORED: &[&str] = &[
    ".git", "node_modules", ".next", "dist", "build", "__pycache__",
    ".venv", ".DS_Store", ".glue-paste",
];

#[derive(Serialize)]
pub struct FileEntry {
    pub name: String,
    #[serde(rename = "type")]
    pub entry_type: String,
    pub path: String,
}

fn is_ignored(name: &str) -> bool {
    IGNORED.contains(&name)
}

fn rel(root: &Path, full: &Path) -> String {
    full.strip_prefix(root)
        .unwrap_or(full)
        .to_string_lossy()
        .replace('\\', "/")
}

fn is_valid_card_id(id: &str) -> bool {
    !id.is_empty()
        && id.len() <= 64
        && id.chars().all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
}

#[tauri::command]
pub fn files_browse(state: State<AppState>, board_id: String, path: Option<String>) -> Result<Vec<FileEntry>, String> {
    let conn = state.db.lock().map_err(map_err)?;
    let board = boards::get(&conn, &board_id)
        .map_err(map_err)?
        .ok_or_else(|| "Board not found".to_string())?;

    let root = PathBuf::from(&board.directory).canonicalize().map_err(map_err)?;
    let sub = path.unwrap_or_default();
    let target = root.join(&sub).canonicalize().unwrap_or_else(|_| root.join(&sub));

    // Prevent directory traversal
    if target != root && !target.starts_with(&root) {
        return Err("Invalid path".to_string());
    }

    let entries = fs::read_dir(&target).map_err(|_| "Cannot read directory".to_string())?;
    let mut result: Vec<FileEntry> = Vec::new();

    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        if is_ignored(&name) { continue; }
        if let Ok(meta) = entry.metadata() {
            let entry_type = if meta.is_dir() { "directory" } else { "file" };
            let full_path = entry.path();
            result.push(FileEntry {
                name,
                entry_type: entry_type.to_string(),
                path: rel(&root, &full_path),
            });
        }
    }

    result.sort_by(|a, b| {
        if a.entry_type != b.entry_type {
            if a.entry_type == "directory" { std::cmp::Ordering::Less } else { std::cmp::Ordering::Greater }
        } else {
            a.name.cmp(&b.name)
        }
    });

    Ok(result)
}

fn walk_tree(dir: &Path, root: &Path, entries: &mut Vec<FileEntry>, depth: usize, max_files: usize) -> bool {
    if depth > 20 || entries.len() >= max_files {
        return true; // truncated
    }
    let read = match fs::read_dir(dir) {
        Ok(r) => r,
        Err(_) => return false,
    };
    for entry in read.flatten() {
        if entries.len() >= max_files {
            return true;
        }
        let name = entry.file_name().to_string_lossy().to_string();
        if is_ignored(&name) { continue; }
        if let Ok(meta) = entry.metadata() {
            let is_dir = meta.is_dir();
            let full = entry.path();
            entries.push(FileEntry {
                name,
                entry_type: if is_dir { "directory" } else { "file" }.to_string(),
                path: rel(root, &full),
            });
            if is_dir {
                if walk_tree(&full, root, entries, depth + 1, max_files) {
                    return true;
                }
            }
        }
    }
    false
}

#[tauri::command]
pub fn files_tree(state: State<AppState>, board_id: String) -> Result<serde_json::Value, String> {
    let conn = state.db.lock().map_err(map_err)?;
    let board = boards::get(&conn, &board_id)
        .map_err(map_err)?
        .ok_or_else(|| "Board not found".to_string())?;

    let root = PathBuf::from(&board.directory);
    let mut entries: Vec<FileEntry> = Vec::new();
    let truncated = walk_tree(&root, &root, &mut entries, 0, 10000);

    Ok(serde_json::json!({ "entries": entries, "truncated": truncated }))
}

#[tauri::command]
pub fn attachments_list(state: State<AppState>, board_id: String, card_id: String) -> Result<Vec<String>, String> {
    if !is_valid_card_id(&card_id) {
        return Err("Invalid card ID".to_string());
    }
    let conn = state.db.lock().map_err(map_err)?;
    let board = boards::get(&conn, &board_id)
        .map_err(map_err)?
        .ok_or_else(|| "Board not found".to_string())?;

    let root = PathBuf::from(&board.directory);
    let attachments_dir = root.join(".glue-paste").join("attachments").join(&card_id);

    match fs::read_dir(&attachments_dir) {
        Ok(entries) => {
            let files: Vec<String> = entries
                .flatten()
                .map(|e| rel(&root, &e.path()))
                .collect();
            Ok(files)
        }
        Err(_) => Ok(Vec::new()),
    }
}

#[tauri::command]
pub fn attachments_delete_file(
    state: State<AppState>,
    board_id: String,
    card_id: String,
    filename: String,
) -> Result<serde_json::Value, String> {
    if !is_valid_card_id(&card_id) {
        return Err("Invalid card ID".to_string());
    }
    let safe_name = sanitize_filename(&filename);
    if safe_name.is_empty() {
        return Err("Invalid filename".to_string());
    }

    let conn = state.db.lock().map_err(map_err)?;
    let board = boards::get(&conn, &board_id)
        .map_err(map_err)?
        .ok_or_else(|| "Board not found".to_string())?;

    let file_path = PathBuf::from(&board.directory)
        .join(".glue-paste")
        .join("attachments")
        .join(&card_id)
        .join(&safe_name);

    let _ = fs::remove_file(&file_path); // ignore if not found
    Ok(serde_json::json!({ "ok": true }))
}

#[tauri::command]
pub fn attachments_cleanup(
    state: State<AppState>,
    board_id: String,
    card_id: String,
) -> Result<serde_json::Value, String> {
    if !is_valid_card_id(&card_id) {
        return Err("Invalid card ID".to_string());
    }
    let conn = state.db.lock().map_err(map_err)?;
    let board = boards::get(&conn, &board_id)
        .map_err(map_err)?
        .ok_or_else(|| "Board not found".to_string())?;

    let attachments_dir = PathBuf::from(&board.directory)
        .join(".glue-paste")
        .join("attachments")
        .join(&card_id);

    let _ = fs::remove_dir_all(&attachments_dir); // ignore if not found
    Ok(serde_json::json!({ "ok": true }))
}

fn sanitize_filename(name: &str) -> String {
    let base = Path::new(name)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();
    base.replace(['/', '\\', '\0'], "_")
}
